src/E2_economics.py

```diff
@@ -175,3 +177,24 @@ def add_costs_and_total(dict_asset, name, value, total_costs):
 def all_list_in_dict(dict_asset, list):
     boolean = all([name in dict_asset for name in list]) is True
     return boolean
+
+
+def levelized_cost_of_electricity_of_asset(dict_asset):
+    """
+    Calculates the levelized cost of electricity (or generation) of each asset defined in energyProduction.csv
+    Parameters
+    ----------
+    dict_asset
+
+    Returns
+    -------
+
+    """
+    lcoe = {}
+    for asset in dict_asset[ENERGY_PRODUCTION]:
+        lcoe[asset] = (
+            dict_asset[ENERGY_PRODUCTION][asset][ANNUITY_TOTAL][VALUE]
+            / dict_asset[ENERGY_PRODUCTION][asset][TOTAL_FLOW][VALUE]
+        )
```

Comment:
Actually, you do not only want to calculate the LCOE per asset for the production asset - but for all assets. Therefore, you need to loop trough the list of all assets as following: 
```
def lcoe_assets(dict_values):

asset_group_list= [
    ENERGY_CONSUMPTION,
    ENERGY_CONVERSION,
    ENERGY_PRODUCTION,
    ENERGY_PROVIDERS,
    ENERGY_STORAGE]
    for asset_group in asset_group_list:
        for asset in dict_values[asset_group]: 
              lcoe_a = ...
             dict_values[asset_grout][asset].update({LCOE_ASSET: lcoe_a})
```
You should add ` LCOE_ASSET` to the `constants_json_strings.py` file below "processing" for this.

(For energy_consumption this lcoe_a should always be 0)

Reply:
Thanks @smartie2076 :)

Reply:
@smartie2076 What if TOTAL_FLOW of an asset is 0? we skip the update?

Reply:
If `total_flow == 0` we could try setting the `lcoe_a = None`. Not sure if later post-processing will accept this, but we will see ;)

Reply:
yes I set it to NA (will change it to None) but to be honest, I don't understand the current error with the key.

Reply:
Can you post the error message? Is it the one from above that PF tried to explain?

Reply:
Traceback (most recent call last):
  File "mvs_tool.py", line 4, in <module>
    main()
  File "/Users/pastek/Desktop/MVS-RLI/mvs_eland/mvs_eland_tool/local_deploy.py", line 151, in main
    evaluation.evaluate_dict(dict_values, results_main, results_meta)
  File "/Users/pastek/Desktop/MVS-RLI/mvs_eland/src/E0_evaluation.py", line 182, in evaluate_dict
    economics.lcoe_assets(dict_values)
  File "/Users/pastek/Desktop/MVS-RLI/mvs_eland/src/E2_economics.py", line 209, in lcoe_assets
    elif dict_values[asset_group][asset][TOTAL_FLOW][VALUE] == 0.0:
KeyError: 'total_flow'

Reply:
Okay, that means that it can not find that key `total_flow` in the specific asset you are looking at. You 
should

 a) do `print(dict_values[asset_group][asset].keys())` and check for the `total_flow` attribute 
or 
b) check `json_with_results.json` if the key exists in the attribute the loop is in `print(asset group, asset)`

Reply:
so basically there is not "total_flow" in ENERGY_PROVIDERS and ENERGY_STORAGE
I commented them out for the moment. What do you think?

Reply:
It makes sense that it does not work for `energyProviders`. However, you should add a second loop below that asseses the `energyStorages`: That has a cost connected.

You should add the annuities of the sub-assets (capacity, input/output power) to one annuity for the storage, and devide it by the timeseries stored in output power. That will give us the total levelized cost to dispatch one kWh from the batteries.